File: TP's/TP 4/Main.java
import java.io.File;
import java.io.FileNotFoundException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;

class Tp4 {
    private String appID;
    private String name;
    private String releaseDate;
    private String compradores;
    private String preco;
    private String score;
    private String userScore;
    private String nConquistas;
    private String supportedLanguages;
    private String publishers;
    private String developers;
    private String categorias;
    private String generos;
    private String tags;

    public Tp4() {}

    public void setAppID(String appID) { this.appID = appID.trim(); }
    public void setName(String name) { this.name = name.trim(); }
    public void setReleaseDate(String releaseDate) { this.releaseDate = releaseDate.trim(); }
    public void setCompradores(String compradores) { this.compradores = compradores.trim(); }
    public void setPreco(String preco) { this.preco = preco.trim(); }
    public void setScore(String score) { this.score = score.trim(); }
    public void setUserScore(String userScore) { this.userScore = userScore.trim(); }
    public void setnConquistas(String nConquistas) { this.nConquistas = nConquistas.trim(); }
    public void setSupportedLanguages(String supportedLanguages) { this.supportedLanguages = supportedLanguages.trim(); }
    public void setPublishers(String publishers) { this.publishers = publishers.trim(); }
    public void setDevelopers(String developers) { this.developers = developers.trim(); }
    public void setCategorias(String categorias) { this.categorias = categorias.trim(); }
    public void setGeneros(String generos) { this.generos = generos.trim(); }
    public void setTags(String tags) { this.tags = tags.trim(); }

    public String getAppID() { return appID; }

    @Override
    public String toString() {
        String dataFormatada = releaseDate;
        try {
            SimpleDateFormat entrada = new SimpleDateFormat("MMM dd, yyyy", Locale.ENGLISH);
            SimpleDateFormat saida = new SimpleDateFormat("dd/MM/yyyy");
            dataFormatada = saida.format(entrada.parse(releaseDate.replace("\"", "")));
        } catch (ParseException e) {}

        return "=> "+ appID + " ## " +
               limpar(name) + " ## " +
               dataFormatada + " ## " +
               limpar(compradores) + " ## " +
               limpar(preco) + " ## " +
               "[" + limparLista(supportedLanguages) + "] ## " +
               limpar(score) + " ## " +
               limpar(userScore) + " ## " +
               limpar(nConquistas) + " ## " +
               "[" + limparLista(publishers) + "] ## " +
               "[" + limparLista(developers) + "] ## " +
               "[" + limparLista(categorias) + "] ## " +
               "[" + limparLista(generos) + "] ## " +
               "[" + limparLista(tags) + "] ##";
    }

    private String limpar(String campo) {
        if (campo == null) return "";
        return campo.replace("\"", "").replace("[", "").replace("]", "").trim();
    }

    private String limparLista(String campo) {
        campo = limpar(campo);
        String[] partes = campo.split(",");
        String resultado = "";
        for (int i = 0; i < partes.length; i++) {
            String item = partes[i].trim();
            if (!item.isEmpty()) {
                if (!resultado.isEmpty()) resultado += ", ";
                resultado += item;
            }
        }
        return resultado;
    }
}

public class Main {
    public static void main(String[] args) {
        String path = "/tmp/games.csv";
        List<Tp4> jogos = new ArrayList<>();

        try (Scanner scanner = new Scanner(new File(path))) {
            if (scanner.hasNextLine()) scanner.nextLine();

            while (scanner.hasNextLine()) {
                String linha = scanner.nextLine();
                String[] campos = linha.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)", -1);

                Tp4 jogo = new Tp4();
                jogo.setAppID(campos[0]);
                jogo.setName(campos[1]);
                jogo.setReleaseDate(campos[2]);
                jogo.setCompradores(campos[3]);
                jogo.setPreco(campos[4]);
                jogo.setSupportedLanguages(campos[5]);
                jogo.setScore(campos[6]);
                jogo.setUserScore(campos[7]);
                jogo.setnConquistas(campos[8]);
                jogo.setPublishers(campos[9]);
                jogo.setDevelopers(campos[10]);
                jogo.setCategorias(campos[11]);
                jogo.setGeneros(campos[12]);
                jogo.setTags(campos[13]);

                jogos.add(jogo);
            }
        } catch (FileNotFoundException e) {
            System.out.println("Arquivo não encontrado: " + e.getMessage());
            return;
        }

        Scanner input = new Scanner(System.in);
        String idBusca;
        while (true) {
            idBusca = input.nextLine().trim();
            if (idBusca.equalsIgnoreCase("Fim")) break;

            boolean encontrado = false;
            for (Tp4 jogo : jogos) {
                if (idBusca.equals(jogo.getAppID())) {
                    System.out.println(jogo);
                    encontrado = true;
                }
            }
            if (!encontrado) System.out.println("Jogo com ID " + idBusca + " não encontrado.");
        }
    }
}

File: TP's/TP 4/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char *appID;
    char *name;
    char *releaseDate;
    char *compradores;
    char *preco;
    char *score;
    char *userScore;
    char *nConquistas;
    char *supportedLanguages;
    char *publishers;
    char *developers;
    char *categorias;
    char *generos;
    char *tags;
} Tp4;

char* str_alloc(const char* s) {
    if (!s) return NULL;
    char *p = malloc(strlen(s) + 1);
    strcpy(p, s);
    return p;
}

void limpar(char* campo) {
    if (!campo) return;
    char *src = campo, *dst = campo;
    while (*src) {
        if (*src != '"' && *src != '[' && *src != ']') *dst++ = *src;
        src++;
    }
    *dst = '\0';
}

void formatList(char* campo) {
    if (!campo) return;
    char temp[1024], resultado[1024] = "";
    strcpy(temp, campo);
    limpar(temp);

    char *token = strtok(temp, ",");
    int primeiro = 1;
    while (token) {
        while (*token == ' ') token++;
        if (!primeiro) strcat(resultado, ", ");
        strcat(resultado, token);
        primeiro = 0;
        token = strtok(NULL, ",");
    }
    strcpy(campo, resultado);
}

void formatDate(const char* input, char* output) {
    const char* meses[] = {"Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"};
    int mes = 0, dia = 0, ano = 0;
    char mesStr[4];
    if (sscanf(input, "\"%3s %d, %d\"", mesStr, &dia, &ano) == 3) {
        for (mes = 0; mes < 12; mes++) {
            if (strcmp(mesStr, meses[mes]) == 0) break;
        }
        sprintf(output, "%02d/%02d/%04d", dia, mes+1, ano);
    } else {
        strcpy(output, input);
    }
}

void printTp4(Tp4 *jogo) {
    char dataFormatada[20];
    formatDate(jogo->releaseDate, dataFormatada);

    formatList(jogo->supportedLanguages);
    formatList(jogo->publishers);
    formatList(jogo->developers);
    formatList(jogo->categorias);
    formatList(jogo->generos);
    formatList(jogo->tags);

    limpar(jogo->name);
    limpar(jogo->compradores);
    limpar(jogo->preco);
    limpar(jogo->score);
    limpar(jogo->userScore);
    limpar(jogo->nConquistas);

    printf("=> %s ## %s ## %s ## %s ## %s ## [%s] ## %s ## %s ## %s ## [%s] ## [%s] ## [%s] ## [%s] ## [%s] ## [%s] ##\n",
           jogo->appID,
           jogo->name,
           dataFormatada,
           jogo->compradores,
           jogo->preco,
           jogo->supportedLanguages,
           jogo->score,
           jogo->userScore,
           jogo->nConquistas,
           jogo->publishers,
           jogo->developers,
           jogo->categorias,
           jogo->generos,
           jogo->tags);
}

void freeTp4(Tp4 *jogo) {
    free(jogo->appID);
    free(jogo->name);
    free(jogo->releaseDate);
    free(jogo->compradores);
    free(jogo->preco);
    free(jogo->score);
    free(jogo->userScore);
    free(jogo->nConquistas);
    free(jogo->supportedLanguages);
    free(jogo->publishers);
    free(jogo->developers);
    free(jogo->categorias);
    free(jogo->generos);
    free(jogo->tags);
    free(jogo);
}

int main() {
    const char *path = "/tmp/games.csv";
    FILE *fp = fopen(path, "r");
    if (!fp) {
        printf("Arquivo não encontrado.\n");
        return 1;
    }

    char linha[4096];
    if (!fgets(linha, sizeof(linha), fp)) {
        fclose(fp);
        return 1;
    }

    char idBusca[128];
    while (1) {
        if (!fgets(idBusca, sizeof(idBusca), stdin)) break;
        idBusca[strcspn(idBusca, "\n")] = 0;
        if (strcmp(idBusca, "Fim") == 0) break;

        rewind(fp);
        fgets(linha, sizeof(linha), fp);
        int encontrado = 0;
        while (fgets(linha, sizeof(linha), fp)) {
            Tp4 *jogo = malloc(sizeof(Tp4));
            memset(jogo, 0, sizeof(Tp4));

            char *campos[14];
            char *token = strtok(linha, ",");
            int i = 0;
            while (token && i < 14) {
                campos[i++] = token;
                token = strtok(NULL, ",");
            }

            jogo->appID = str_alloc(campos[0]);
            if (strcmp(jogo->appID, idBusca) == 0) {
                jogo->name = str_alloc(campos[1]);
                jogo->releaseDate = str_alloc(campos[2]);
                jogo->compradores = str_alloc(campos[3]);
                jogo->preco = str_alloc(campos[4]);
                jogo->supportedLanguages = str_alloc(campos[5]);
                jogo->score = str_alloc(campos[6]);
                jogo->userScore = str_alloc(campos[7]);
                jogo->nConquistas = str_alloc(campos[8]);
                jogo->publishers = str_alloc(campos[9]);
                jogo->developers = str_alloc(campos[10]);
                jogo->categorias = str_alloc(campos[11]);
                jogo->generos = str_alloc(campos[12]);
                jogo->tags = str_alloc(campos[13]);

                printTp4(jogo);
                encontrado = 1;
                freeTp4(jogo);
                break;
            }
            free(jogo->appID);
            free(jogo);
        }
    }

    fclose(fp);
    return 0;
}
